process BAMBU {
    label 'process_medium'

    conda "conda-forge::r-base=4.0.3 bioconda::bioconductor-bambu=3.0.8 bioconda::bioconductor-bsgenome=1.66.0"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bioconductor-bambu:3.0.8--r42hc247a5b_0' :
        'quay.io/biocontainers/bioconductor-bambu:3.0.8--r42hc247a5b_0' }"

    input:
    tuple path(fasta), path(gtf)
    path bams

    output:
    path "counts_gene.txt"         , emit: ch_gene_counts
    path "counts_transcript.txt"   , emit: ch_transcript_counts
    path "extended_annotations.gtf", emit: extended_gtf
    path "versions.yml"            , emit: versions
    path "bambu_ndr.csv"           , emit: ndr

    when:
    task.ext.when == null || task.ext.when

    script:
    """
    run_bambu.r \\
        --tag=. \\
        --ncore=$task.cpus \\
        --annotation=$gtf \\
        --fasta=$fasta \\
        $bams

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        bioconductor-bambu: \$(Rscript -e "library(bambu); cat(as.character(packageVersion('bambu')))")
        bioconductor-bsgenome: \$(Rscript -e "library(BSgenome); cat(as.character(packageVersion('BSgenome')))")
    END_VERSIONS

    sed -i 's/*/./g' extended_annotations.gtf
    """
}
